File: examples/exn_bar2m.py
# example exn_bar2m
# --------------------------------------------------------------------------
# PURPOSE
#    Analysis of a plane truss considering material nonlinearity.
# --------------------------------------------------------------------------

import numpy as np
import calfem.core as cfc
import calfem.utils as cfu
import calfem.vis_mpl as cfv

edof = np.array([
    [1, 2, 5, 6], 
    [5, 6, 7, 8], 
    [3, 4, 5, 6]
])

bc = np.array([1, 2, 3, 4, 7, 8])

ex = np.array([
    [0.0, 1.6], 
    [1.6, 1.6], 
    [0.0, 1.6]
])

ey = np.array([[0.0, 0.0], [0.0, 1.2], [1.2, 0.0]])

E = np.array([200e9, 200e9, 200e9])
A = np.array([6.0e-4, 3.0e-4, 10.0e-4])
SY = 400e6
Ns = (SY * A).reshape(-1, 1)

dp = 4e3

incr = 100

a = np.zeros((8, 1))
r = np.zeros((8, 1))
es = np.zeros((3, 1))

plbar = 0
pl = np.array([0.0, 0.0])

# Forward Euler increamental solution

for i in range(incr):
    K = np.zeros((8, 8))
    df = np.zeros((8, 1))
    df[5] = -dp

    # Create and assemble element tangent stiffness matrix

    for j in range(3):
        ep = np.array([E[j], A[j]])
        Ke = cfc.bar2e(ex[j, :], ey[j, :], ep)
        K = cfc.assem(edof[j, :], K, Ke)

    # Stop iteration if determinant det(Kr) <= 0

    fdof = np.setdiff1d(np.arange(1, 9), bc) - 1
    Kr = K[np.ix_(fdof, fdof)]
    if np.linalg.det(Kr) <= 0:
        print("Determinant zero after increment ", i)
        break

    # Solve for the displacement increment and determine total displacements

    da, dr = cfc.solveq(K, df, bc)
    a += da
    r += dr

    # Determine normal forces in elements

    ded = cfc.extract_ed(edof, da)
    des = np.zeros((3, 1))

    for j in range(3):
        ep = np.array([E[j], A[j]])
        desj = cfc.bar2s(ex[j, :], ey[j, :], ep, ded[j, :])
        des[j, 0] = desj[0]

    es += des

    for j in range(3):
        if abs(es[j, 0]) >= Ns[j]:
            E[j] = 0

    # Determine if the stress in a bar has reached the yield stress

    newplbar = np.sum(abs(es) > Ns)

    if newplbar > plbar:
        plbar = newplbar
        print(
            plbar, "plastic elements for increment ", i + 1, " at load = ", (i + 1) * dp
        )

    # Save variables for curve plotting

    pl = np.vstack((pl, np.array([-a[5, 0], (i + 1) * dp])))

# Plot force-displacement relation

cfv.figure(1, fig_size=(7, 4))
cfv.plt.plot(pl[:, 0], pl[:, 1])
cfv.plt.xlabel("Displacement")
cfv.plt.ylabel("Force")
cfv.show_and_wait()

File: examples/exn_bar2m.m
% example exn_bar2m
%----------------------------------------------------------------
% PURPOSE 
%    Analysis of a plane truss with material nonlinearity.
%----------------------------------------------------------------

% LAST MODIFIED: O Dahlblom  2015-12-02
% Copyright (c)  Division of Structural Mechanics and
%                Division of Solid Mechanics.
%                Lund University
%----------------------------------------------------------------

echo off

edof=[1 1 2 5 6;
      2 5 6 7 8;
      3 3 4 5 6]

bc=[1 0; 2 0; 3 0; 4 0; 7 0; 8 0];

ex=[0 1.6; 1.6 1.6; 0 1.6]
ey=[0 0; 0 1.2; 1.2 0]

Em=200.0e9;
E=ones(3,1)*Em;
A1=6.0e-4; A2=3.0e-4; A3=10.0e-4;
A=[A1 A2 A3]'
SY=400.0e6;
Ns=SY*A

dp=4.0e3

incr=100;

a=zeros(8,1);
r=zeros(8,1);
es=zeros(3,1);

plbar=0;
pl(1,:)=[0 0];

% Forward Euler incremental solution

for i=1:incr
   K=zeros(8);
   df=zeros(8,1);
   df(6)=-dp;
  
   % Create and assemble element tangent stiffness matrices
   for j=1:3  
      ep=[E(j),A(j)];
      Ke=bar2e(ex(j,:),ey(j,:),ep);
      K=assem(edof(j,:),K,Ke);
   end;
      
   % Stop iteration if determinant det(Kr)<=0
   Kr=red(K,bc(:,1));
   if det(Kr)<=0 
      disp(['Determinant zero after increment ',num2str(i-1)])
      break;
  end;
   
   % Solve for the displacement increment and determine total displacements
   [da,dr]=solveq(K,df,bc);
   a=a+da;
   r=r+dr;
   
   % Determine normal forces in elements
   ded=extract_ed(edof,da);
   for j=1:3
     ep=[E(j),A(j)];
     desj=bar2s(ex(j,:),ey(j,:),ep,ded(j,:));
     des(j,1)=desj(1);
   end;
   es=es+des;
   for j=1:3
     E(j)=Em; if abs(es(j))>=Ns(j);  E(j)=0; end 
   end; 
   
   % Determine if the stress in a bar has reached the yield stress
   newplbar=sum(abs(es)>Ns);
   if newplbar > plbar
     plbar=newplbar;
     disp([num2str(plbar),' plastic elements for increment ',num2str(i), ...
         ' at load = ', num2str(i*dp)])
     es
   end;
   
   % Save variables for curve plotting
   pl(i+1,:)=[-a(6),i*dp];
 
end;
   % Plot force-displacement relation
   figure(1)
   plot(pl(:,1),pl(:,2),'-');
